uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/BAO/LeaveBalanceChange.php

```diff
@@ -410,94 +411,98 @@ public static function getAmountForTOILRequest($toilRequestID) {
    *
    * @return int The number of records created
    */
-  public static function createExpirationRecords() {
+  public static function createExpiryRecords() {
     $numberOfRecordsCreated = 0;
 
-    $tableName = self::getTableName();
-    $query = "
-      SELECT balance_to_expire.*
-      FROM {$tableName} balance_to_expire
-      LEFT JOIN {$tableName} expired_balance_change
-             ON balance_to_expire.id = expired_balance_change.expired_balance_change_id
-      WHERE balance_to_expire.expiry_date IS NOT NULL AND
-            balance_to_expire.expiry_date < CURDATE() AND
-            balance_to_expire.expired_balance_change_id IS NULL AND
-            expired_balance_change.id IS NULL
-      ORDER BY balance_to_expire.source_id ASC, balance_to_expire.expiry_date ASC
-    ";
-
-    $startDate = null;
-    $sourceID = null;
-    $sourceType = null;
-
-    $dao = CRM_Core_DAO::executeQuery($query, [], true, self::class);
-    while($dao->fetch()) {
-      if($dao->source_id != $sourceID && $dao->source_type != $sourceType) {
-        $startDate = null;
-        $sourceID = $dao->source_id;
-        $sourceType = $dao->source_type;
+    $balanceChangesToExpire = self::getBalanceChangesToExpire();
+    $datesOverlappingBalanceChangesToExpire = self::getDatesOverlappingBalanceChangesToExpire($balanceChangesToExpire);
+
+    foreach($balanceChangesToExpire as $balanceChangeToExpire) {
+      $remainingAmount = $balanceChangeToExpire['amount'];
+
+      foreach($datesOverlappingBalanceChangesToExpire as $i => $date) {
+        if($date['date'] >= $balanceChangeToExpire['start_date'] && $date['date'] <= $balanceChangeToExpire['expiry_date']) {
+          if($remainingAmount >= abs($date['amount'])) {
+            // Date already deducted, so new we remove it from the
+            // array so it won't be deducted again from another
+            // balance change
+            unset($datesOverlappingBalanceChangesToExpire[$i]);
+            $remainingAmount += $date['amount'];
```

Comment:
There will be some instances when `$remainingAmount` would have become zero and the dates loop will still run for all the remaining dates even though the this condition `if($remainingAmount >= abs($date['amount']))` is there, it will still run through the whole date nonetheless to check that condition. We can short circuit by checking if `$remainingAmount = 0` and then escape the loop altogether and move to the next `$balanceChangeToExpire` in line.